tests/002-send-receive.cpp

```diff
@@ -269,6 +351,206 @@ namespace oxen::quic::test
+
+        CHECK(done.wait_for(5s) == std::future_status::ready);
+        std::lock_guard lock{mut};
+        CHECK(good_responses == num_requests);
+        CHECK(responses == good_responses);
+    }
+
+    TEST_CASE("002 - BParser huge requests", "[002][bparser][huge]")
+    {
+        Network test_net{};
+
+        auto [client_tls, server_tls] = defaults::tls_creds_from_ed_keys();
+
+        Address server_local{};
+        Address client_local{};
+
+        static constexpr int num_requests = 10;
+
+        std::mutex mut;
+        std::promise<void> done_prom;
+        auto done = done_prom.get_future();
+        int responses = 0, good_responses = 0;
+
+        SECTION("Huge but not too huge")
+        {
+            std::string req_msg(9'000'000, 'a');
+            constexpr auto res_msg = "oh look some a's"sv;
+
+            auto server_handler = [&](message msg) {
+                if (msg)
+                {
+                    if (msg.body() == req_msg)
+                        msg.respond(res_msg);
+                    else
+                        msg.respond("where are all the a's?!");
+                }
+            };
+
+            auto client_reply_handler = [&](message msg) {
+                if (msg)
+                {
+                    std::lock_guard lock{mut};
+                    responses++;
+                    log::debug(log_cat, "Client bparser received response {}: {}", responses, msg.view());
+                    if (msg.body() == res_msg)
+                        good_responses++;
+                    if (responses == num_requests)
+                        done_prom.set_value();
+                }
+                else
+                {
+                    log::debug(log_cat, "got back a failed message response");
+                }
+            };
+
+            stream_constructor_callback server_constructor = [&](Connection& c, Endpoint& e, std::optional<int64_t>) {
+                auto s = e.make_shared<BTRequestStream>(c, e);
+                s->register_command("test_endpoint"s, server_handler);
+                return s;
+            };
+
+            auto server_endpoint = test_net.endpoint(server_local);
+            REQUIRE_NOTHROW(server_endpoint->listen(server_tls, server_constructor));
+
+            RemoteAddress client_remote{defaults::SERVER_PUBKEY, "127.0.0.1"s, server_endpoint->local().port()};
+
+            auto client_endpoint = test_net.endpoint(client_local);
+            auto conn_interface = client_endpoint->connect(client_remote, client_tls);
+
+            std::shared_ptr<BTRequestStream> client_bp = conn_interface->get_new_stream<BTRequestStream>();
+
+            for (int i = 0; i < num_requests; i++)
+            {
+                client_bp->command("test_endpoint"s, req_msg, client_reply_handler);
+            }
+
+            CHECK(done.wait_for(5s) == std::future_status::ready);
+            std::lock_guard lock{mut};
+            CHECK(good_responses == num_requests);
```

Comment:
Okay so I dug into why this is failing literally all the CI builds. I separated the test cases to look into the one that was failing specifically, pushed it to [my branch](https://github.com/dr7ana/libquic/blob/f20358f345e463c773c41452972a7a7764ca0f6f/tests/002-send-receive.cpp#L432) to keep testnet tracking the pr void-listen

[This is the offending line](https://github.com/dr7ana/libquic/blob/f20358f345e463c773c41452972a7a7764ca0f6f/src/connection.cpp#L910) that slows down the test to the point of failure, and downgrading it's log-level to trace makes the test work as you wrote it.

Though what's interesting is if you leave that line as a debug log, put [this line](https://github.com/dr7ana/libquic/blob/f20358f345e463c773c41452972a7a7764ca0f6f/tests/002-send-receive.cpp#L513) instead of the `done.wait_for(5s)` call, and run the test suit with debug or trace logging, it loops forever (both with and without the mutexes)